subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultWatchingVirtualFileSystem.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.vfs.impl;
+
+import com.sun.nio.file.SensitivityWatchEventModifier;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.vfs.WatchingVirtualFileSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+public class DefaultWatchingVirtualFileSystem extends AbstractDelegatingVirtualFileSystem implements WatchingVirtualFileSystem, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWatchingVirtualFileSystem.class);
+
+    private final Predicate<String> watchFilter;
+
+    private WatchService watchService;
+
+    public DefaultWatchingVirtualFileSystem(
+        AbstractVirtualFileSystem delegate,
+        Predicate<String> watchFilter
+    ) {
+        super(delegate);
+        this.watchFilter = watchFilter;
+    }
+
+    @Override
+    public void startWatching() {
+        if (watchService != null) {
+            throw new IllegalStateException("Watch service already started");
+        }
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+        Set<File> visited = new HashSet<>();
+        getRoot().visitCompleteSnapshots(snapshot -> {
+            String absolutePath = snapshot.getAbsolutePath();
+            if (!watchFilter.test(absolutePath)) {
+                return;
+            }
+            File file = new File(absolutePath);
+            try {
+                if (snapshot.getType() == FileType.Directory) {
```

Comment:
Can you add some comment here explaining why we watch the direct parent of any complete snapshot? Some explanation how the watching is expected to work would be great.

Reply:
Added some notes.